README.md
# erpPCA

R port of MATLAB functions erpPCA.m and others created by Jürgen Kayser. The original MATLAB functions can be freely downloaded from http://psychophysiology.cpmc.columbia.edu/mmedia/Kayser2003a/Appendix.html

## Scope

This package implements only the PCA + Varimax rotation workflow advocated by Kayser and Tenke (2003, 2005, 2006) for Event-Related Potential (ERP) data, and is *not* a general approach to PCA nor to factor analysis.

A more flexible factor analysis approach to ERP data can be found in the ERP PCA Toolkit by Joseph Dien (Dien, 2010a), which is available only in MATLAB. Perhaps a future version of this package will implement some aspects of this more flexible approach, namely:

* The ability to use different rotations (especially Promax, as recommended by Dien (2010b, Dien et al., 2007)

* The ability to extract and rotate only a user-defined number of factors.

The R ecosystem already provides plenty of PCA and factor analyses packages that can probably be co-opted for these purposes.

## References

* Dien, J., Khoe, W., & Mangun, G. R. (2007). Evaluation of PCA and ICA of simulated ERPs: Promax vs. Infomax rotations. Human brain mapping, 28(8), 742-763.

* Dien, J. (2010a). The ERP PCA Toolkit: An open source program for advanced statistical analysis of event-related potential data. Journal of neuroscience methods, 187(1), 138-145.

* Dien, J. (2010b). Evaluating two‐step PCA of ERP data with geomin, infomax, oblimin, promax, and varimax rotations. Psychophysiology, 47(1), 170-183.

* Kayser, J., & Tenke, C. E. (2003). Optimizing PCA methodology for ERP component identification and measurement: theoretical rationale and empirical evaluation. Clinical Neurophysiology, 114(12), 2307-2325.

* Kayser, J., & Tenke, C. E. (2005). Trusting in or breaking with convention: towards a renaissance of principal components analysis in electrophysiology. Clinical Neurophysiology, 116(8), 1747-1753.

* Kayser, J., & Tenke, C. E. (2006). Consensus on PCA for ERP data, and sensibility of unrestricted solutions. Clinical Neurophysiology, 117(3), 703-707.



NEWS.md
# Version 1.0 (May 2015)

* Full port of erpPCA.m and related MATLAB functions from Jurgen Kayser completed.

* There are a couple of options in the package that go beyond the simple porting of the original MATLAB functions. These include
    * Decompose the data matrix with eigen or svd (original functions use eigen)
    * Two options for estimating the rank of the svd matrix (original, which uses svd of correlation matrix, and qr, which does not call svd, and therefore is probably more computationally efficient.). The options are not interchangeable, though.
    * Two options for Varimax rotation. The original solution uses the dedicated function DoVarimax4M, but the user may also use the base R varimax function.

R/DoVarimax4M.R
#' @title Performs Varimax rotation of principal components or factors.
#' 
#' @description This function performs a Varimax rotation on factor loadings 
#' from an eigenvalue or singular value decomposition. According to the original
#' MATLAB code by Jurgen Kayser, it ``emulates algorithms described by Harman
#' (1967, pp. 304-308) as implemented in BMDP-4M (Dixon, 1992, pp. 602-603).''
#' 
#' @param X A matrix. Should be the loadings matrix from eigenvalue or 
#'   singular value decomposition.
#'   
#' @param maxit Integer. Maximum number of iterations allowed before accepting
#'  a solution to the rotation. Defaults to 100 iterations.
#'  
#' @param tol Convergence criterion. Defaults to 0.0001.
#' 
#' @param normalize Logical scalar. Should Kaiser normalization be used? 
#'   Defaults to TRUE.
#'   
#' @param verbose Logical scalar. Should text output on the attempted rotation 
#'   be displayed? Defaults to FALSE.
#'   
#' @return List containing 
#' \item{Y}{A matrix. Rotated loadings from \code{X}.}
#' \item{G}{A matrix. History of attempted solutions for the rotation problem.}
#'   
#' @seealso \code{\link[stats]{varimax}} from package \code{stats} and
#'   \code{\link[GPArotation]{Varimax}} from package \code{GPArotation} for
#'   other implementations of Varimax rotations.
#'   
#' @export
#'   
#' @examples
#' ## from the help of stats::varimax, using factor analysis
#' fa <- factanal( ~., 2, data = swiss)
#' stats::varimax(loadings(fa), normalize = FALSE)
#' DoVarimax4M(loadings(fa), normalize = FALSE)$Y
#' 
#' ## using the iris dataset and principal components analysis
#' iris.pca <- prcomp(iris[, 1:4])
#' stats::varimax(iris.pca$rotation)
#' DoVarimax4M(iris.pca$rotation)$Y
DoVarimax4M <- function(X, maxit = 100L, tol = 1e-4, normalize = TRUE, 
                        verbose = FALSE) {
  
  # Auxiliary function definitions
  # ------------------------------
  # repmat (emulate repmat from matlab)
  # ComputeSimplicityCriterion
  # ------------------------------
  repmat <- function(a, n, m) {
    kronecker(matrix(1, n, m), a)
  }
  #
  ComputeSimplicityCriterion <- function(Y) {
    g <- 0
    ncol.Y <- ncol(Y)
    nrow.Y <- nrow(Y)
    for (i in 1:ncol.Y) {
      for (j in 1:ncol.Y) {
        if (i != j) {
          g <- g + sum( (Y[, i]^2) * (Y[, j]^2) ) -
            ((1/nrow.Y) * (sum(Y[, i]^2) * sum(Y[, j]^2)))
        }
      }
    }
    G <- g
    G
  } 
  
  # Varimax4M starts here.
  if (verbose) {
    cat("--------- Varimax Rotation (4M) -------------\n")
  }
  p <- nrow(X)
  m <- ncol(X)
  
  if (verbose) {
    cat("Matrix rows:                     ", p,     "\n")
    cat("Matrix columns:                  ", m,     "\n")
    cat("Max. # of iterations:            ", maxit, "\n")
    cat("Convergence criterion:           ", tol,   "\n")
  }  
  if (normalize) {
    if (verbose) {
      # Use Kaiser's normalization across the rows
      cat("Kaiser's Normalization:          ", "Yes",   "\n")
    }
    h <- sqrt(colSums(t(X)^2))   # communality column vector
    H <- repmat(h, 1, m)  # communality normalization matrix
    Y <- X / H               # normalize X by rows
    Y[is.nan(Y)] <- X[is.nan(Y)]
  } else {
    if (verbose) {
      cat("Kaiser's Normalization:          ", "No",   "\n")
    }
    Y <- X
  }
  
  # Compute rotation criterion for input matrix
  # SimplicityG function
  g <- ComputeSimplicityCriterion(Y)
  
  it <- 0
  G <- matrix(0, nrow = maxit, ncol = 3)
  G[1, ] <- c(it, g, tol)
  Gold <- g    # previous simplicity criterion
  YY <- t(Y)   # rotated matrix at begining of current iteration
  if (verbose) {
    cat("     #    SimplicityCriterion     Convergence\n")
    cat(sprintf("%6d     %18.8f %14.8f", it, g, tol))
    cat("\n")
  }
  for (it in 1:maxit) {
    for (i in 1:(m - 1)) {
      for (j in (i + 1):m) {
        # computes the rotation angle phi as the angle in the complex plane 
        # (Park, 2003):
        t.park <- Arg(
          sum(complex(real = Y[, i], imaginary = Y[, j])^4) / p -
            (sum(complex(real = Y[, i], imaginary = Y[, j])^2) / p)^2
        ) / 4
        
        # rotate the two vectors
        XY <- matrix(c(Y[,i], Y[, j]), ncol = 2, byrow = F) %*% matrix(
          c(cos(t.park), -sin(t.park), sin(t.park), cos(t.park)), nrow = 2, 
          byrow = T
        )
        
        # replace the two columns in the matrix
        Y[, i] <- XY[, 1]
        Y[, j] <- XY[, 2]      
      }
    }
    # Compute rotation criterion for the iteration
    g <- ComputeSimplicityCriterion(Y)
    if (verbose) {
      cat(sprintf("%6d     %18.8f %14.8f", it, g, Gold - g))
      cat("\n")
    }
    if ((Gold - g) < tol) {
      if (Gold < g) {     # if the previous solution was better
        Y <- YY           # report the previous one
      } else {                  
        G[it + 1, ] <- c(it, g, (Gold - g))
      }
      break
    }    
    
    YY <- Y
    G[it + 1, ] = c(it, g, (Gold - g))
    Gold <- g
  }
  
  if (normalize) {
    Y <- Y * H             # reverse Kaiser's normalization
  }
  if (verbose) {
    cat("---------------------------------------------\n")
  }
  output <- list(Y = Y, G = G[1:it, ])
  output
}


R/erpPCA-package.R
#' @title Unrestricted, unstandardized covariance-based PCA with Varimax 
#' rotation for Event-Related Potentials (ERPs).
#' 
#' @description This package implements the set of recommendations laid out by
#' Jurgen Kayser and Craig E. Tenke on a series of publications (see references 
#' below) on how to optimize the decomposition of Event-Related Potential (ERP)
#' data into useful latent factors. These authors propose a PCA decomposition of
#' an ERP data matrix based on its unstandardized covariance matrix, followed by
#' a Varimax rotation of all of its principal components, in an attempt to find 
#' simpler latent factors that may correspond to the event-related potentials 
#' that researchers using EEG/MEG are interesting in measuring.
#' 
#' The proposed workflow involves the decomposition of the ERP matrix in its 
#' principal components using the unstandardized covariance matrices of the 
#' original ERP data matrix followed by a Varimax rotation of all of the 
#' principal components (what Kayser and Tenke call an ``unrestricted'' 
#' solution).
#'
#' @references Kayser, J., & Tenke, C. E. (2003). Optimizing PCA methodology for
#'   ERP component identification and measurement: theoretical rationale and 
#'   empirical evaluation. Clinical Neurophysiology, 114(12), 2307-2325.
#'   
#' @references Kayser, J., & Tenke, C. E. (2005). Trusting in or breaking with 
#'   convention: towards a renaissance of principal components analysis in 
#'   electrophysiology. Clinical Neurophysiology, 116(8), 1747-1753.
#'   
#' @references Kayser, J., & Tenke, C. E. (2006). Consensus on PCA for ERP data,
#'   and sensibility of unrestricted solutions. Clinical Neurophysiology, 
#'   117(3), 703-707.
#'   
#' @source This package is a port of the original MATLAB functions created by
#'   Jurgen Kayser, which can be freely downloaded from 
#'   \url{http://psychophysiology.cpmc.columbia.edu/mmedia/Kayser2003a/Appendix.html}
#'   
#'   
#'
#' @name erpPCA-package
#' @docType package
#' @author Diogo Almeida \email{diogo@@nyu.edu}
#' @keywords package
NULL
#' @title Data from tutorial on PCA applications to Event-Related Potentials
#'   data by Dien and Frishkoff.
#'   
#' @description This dataset is used by Dien and Frishkoff on their book chapter
#'   ``Principal Components Analysis of ERP Data'' (see reference below) in
#'   their a step by step tutorial.
#' 
#' @name dien2005
#' @docType data
#' @usage dien2005
#' @format A data frame with 10 columns and 20 rows
#' @section Variables:
#'
#' \itemize{ 
#' \item \code{Ss}: Subject codes.
#' \item \code{Cs}: Condition codes. 
#' \item \code{t1-t6}: Time samples 1 to 6.
#' }
#'
#' @references Dien, J., & Frishkoff, C. A. (2005). Principal Components 
#'   Analysis of ERP Data. In Handy, T. (Ed.) Event-related Potentials: A
#'   Methods Handbook (pp. 189-207). Cambridge, MA: MIT Press.
NULL

R/erpPCA.R
#' @title Performs unrestricted, unstandardized covariance-based PCA followed by
#'   Varimax rotation
#' 
#' @description This function implements a PCA decomposition of a data matrix 
#' based on its unstandardized covariance matrix, followed by a Varimax rotation
#' of all of its principal components, in an attempt to find simpler latent 
#' factors.
#' 
#' @details This workflow of matrix decomposition followed by Varimax rotation 
#'   has been advocated for electrophysiological studies on humans using the 
#'   Event-Related Potential (ERP) technique by Jurgen Kayser and Craig E. 
#'   Tenke. These authors suggest that very good results can be obtained by
#'   using (1) unstandardized covariance matrices of the original data matrix
#'   followed by (2) a Varimax rotation of all of the principal components (what
#'   they call an ``unrestricted'' rotation). The arguments are laid out in a
#'   series of papers (see references below)
#'  
#' This function is a R port of the original MATLAB function erpPCA.m created
#' by Jurgen Kayser. According to the original documentation, it implements
#' ``the PCA agorithms used by BMDP-4M (Dixon, 1992) and SPSS 10.0 FACTOR''. The
#' original MATLAB code can be freely downloaded at 
#' \url{http://psychophysiology.cpmc.columbia.edu/mmedia/Kayser2003a/Appendix.html}
#' 
#' @section Temporal vs Spatial PCA:
#' For a temporal decomposition (ie., latent factors will correspond to time
#' courses), the matrix \code{X} should be organized with ERP waveforms of each
#' sensor in rows (cases) and time samples in columns (variables). For spatial
#' decomposition (ie., latent factors will correspond to scalp topographies),
#' the matrix \code{X} should be organized with time samples in rows (cases) and
#' sensors in columns (variables).
#' 
#' @param X A matrix. This is the data matrix to be decomposed into its
#'   principal components.
#'   
#' @param method A string. Method to use for finding principal components of
#'   data matrix. Default is \code{eigen} to match the original MATLAB function,
#'   but \code{svd} is generally recommended for numerical accuracy. Both should
#'   give the same results.
#'   
#' @param rank.method A string. Method to use to estimate the rank of the 
#'   principal component matrix. Default is \code{original}, to match the 
#'   original MATLAB function. Option \code{qr} may be more computationally
#'   efficient, but will give different results from the original method.
#'   
#' @param rotation.fun A string. Function to use for Varimax rotation. Default
#'   is \code{original}, which calls the function \code{\link{DoVarimax4M}} from this package,
#'   to match the original MATLAB function. Option \code{R} will call the base
#'   function \code{varimax}. The outputs of these choices will probably be
#'   numerically different.
#'   
#' @return This function returns a list containing the unrotated principal
#'   component loadings, the Varimax rotated factor loadings in a 
#'   variables-by-factors matrix, the factor scores obtained for the rotated
#'   factors in a cases-by-factors matrix. It also outputs the Eingenvalues
#'   of the covariance matrix and the their explained variance before and after
#'   rotation.
#' 
#' \item{Unrotated}{Matrix containing the unrotated principal components
#' loadings.}
#' 
#' \item{Rotated}{Matrix containing the varimax rotated factor loadings}
#' 
#' \item{Factor.Scores}{Matrix containing the factor scores of the varimax
#' rotated factor loadings.}
#' 
#' \item{Variance}{Matrix containing the Eigenvalues and explained variance
#' before and afte rotation.}
#'   
#' @references Kayser, J., & Tenke, C. E. (2003). Optimizing PCA methodology for
#'   ERP component identification and measurement: theoretical rationale and 
#'   empirical evaluation. Clinical neurophysiology, 114(12), 2307-2325.
#'   
#' @references Kayser, J., & Tenke, C. E. (2005). Trusting in or breaking with 
#'   convention: towards a renaissance of principal components analysis in 
#'   electrophysiology. Clinical Neurophysiology, 116(8), 1747-1753.
#'   
#' @references Kayser, J., & Tenke, C. E. (2006). Consensus on PCA for ERP data,
#'   and sensibility of unrestricted solutions. Clinical Neurophysiology, 
#'   117(3), 703-707.
#'   
#' @seealso \code{\link[stats]{princomp}}, \code{\link[stats]{prcomp}}, 
#'   \code{\link[stats]{factanal}} from package \code{stats} for other basic 
#'   implementations of principal component analysis and factor analysis.
#'   
#' @source
#'   \url{http://psychophysiology.cpmc.columbia.edu/mmedia/Kayser2003a/Appendix.html}
#'   
#' @export
#'    
#' @examples
#' ## using the iris dataset and principal components analysis
#' erpPCA(iris[, 2:4])
#'
#' ## Using the tutorial dataset from Dien and Frishkoff (2005)
#' X <- dien2005[, 3:8]
#' 
#' ## Matrix decomposition methods (eigen and svd) should give identical results
#' ## no matter what the matrix rank estimation method is used.
#' A <- erpPCA(X, method = "eigen", rank.method = "original", 
#'   rotation.fun = "original")
#' B <- erpPCA(X, method = "svd", rank.method = "original", 
#'   rotation.fun = "original")
#' all.equal(A, B, check.attributes = FALSE) 
#' 
#' C <- erpPCA(X, method = "eigen", rank.method = "qr", 
#'   rotation.fun = "original")
#' D <- erpPCA(X, method = "svd", rank.method = "qr", rotation.fun = "original")
#' all.equal(C, D, check.attributes = FALSE)
#' 
#' ## The varimax rotation options will give numerically different results, 
#' ## however. The magnitude of this difference needs to be gauged on a case
#' ## by case basis.
#' E <- erpPCA(X, method = "svd", rank.method = "original", rotation.fun = "R")
#' all.equal(B, E, check.attributes = FALSE)
#' 
erpPCA <- function(X, method = c("eigen", "svd"), 
                   rank.method = c("original", "qr"), 
                   rotation.fun = c("original", "R")) {
##############
# Port the erpPCA.m file from Kayser & Tenke 2003 to R
# Diogo Almeida - 6/21/12
##############

  # Helper function definition
  # ------------------------------
  # repmat (emulate repmat from matlab)
  # ------------------------------
  repmat <- function(a, n, m) {
    kronecker(matrix(1, n, m), a)
  }

  # Input checks
  if (is.list(X)) {
    cat("Your data needs to be input as a matrix. You input a list.\n")
    cat("I'm converting your data to the appropriate format now.\n")
    X <- as.matrix(X)
    if (is.double(X)) {
      cat("Your data is now a matrix!\n")
    }
  }
  # Tests to see if function was called without explicitly selecting arguments
  # Introduces the original defaults.
  if (length(method) > 1){  # user forgot to choose method explicitly
    method <- "eigen"
  }
  if (length(rank.method) > 1){  # user forgot to choose rank.method explicitly
    rank.method <- "original"
  }
  if (length(rotation.fun) > 1){  # user forgot to choose rotation.fun
    rotation.fun <- "original"
  }

  if ((method != "eigen") && (method != "svd")) {
    warning("Unrecognized method. It has to be either `eigen' or `svd'.\nThe function will default to `eigen'.\nPlease check your code.")
    method = "eigen"
  }
  if ((rank.method != "original") && (rank.method != "qr")) {
    warning("Unrecognized rank.method. It has to be either `original' or `qr'.\nThe function will default to `original'.\nPlease check your code.")
    rank.method = "original"
  }
  if ((rotation.fun != "original") && (rotation.fun != "R")) {
    warning("Unrecognized rotation function. It has to be either `original' or `R'.\nThe function will default to `original'.\nPlease check your code.")
    rotation.fun = "original"
  }

  # Calculates the Principal Components
  # The original MATLAB function uses eigen, but we can also use svd
  # 
  D <- cov(X)
  if (method == "eigen") {
    eigenD <- eigen(D)
    EM <- eigenD$vectors
    EV <- eigenD$values
    UL <- EM %*% sqrt(diag(EV))    
  } else {
    # method is svd
    svd.X <- prcomp(X, center = T, scale = F)
    EM <- svd.X$rotation
    EV <- svd.X$sdev^2
    UL <- EM %*% diag(svd.X$sdev)
  }

  # Estimate the number of singular values
  # this is effectively estimating the matrix rank of the correlation matrix
  # The original MATLAB function uses the svd decomposition with a tol value of
  # 1e-4.
  # Package Matrix has function rankMatrix, providing several other options.
  if (rank.method == "original") {
    rk <- sum(svd(cor(X))$d > 1e-4)    
  } else {
    # rank.method is 'qr'
    # This may give different results, but is potentially more computationally
    # efficient.
    rk <- qr(cor(X), tol = 1e-4)$rank
  }

  # Remove the linearly dependent factors and their indices
  u <- sort(EV, decreasing = T)[1:rk]
  LU <- UL[, 1:rk]
  
  # Determine direction of principal components and redirect them if negative
  s <- matrix(1, ncol = rk)
  s[abs(apply(LU, 2, max)) < abs(apply(LU, 2, min))] <- -1
  LU <- LU * repmat(s, dim(LU)[1], 1)

  # Rotate the linearly independent principal components
  switch(rotation.fun,
    original = {RL <- DoVarimax4M(LU)$Y},
    R = {RL <- varimax(LU, normalize = TRUE)$loadings},
    stop("`rotation.fun' not recognized")
  )

  # Sort rotated eigen values in descending order
  EVr <- colSums(RL * RL)
  r   <- sort(EVr, decreasing = T)
  rx  <- order(EVr, decreasing = T)
  LR  <-  RL[, rx]

  # Determine direction of rotated factors and redirect them if negative
  s <- matrix(1, ncol =  ncol(LR))
  s[abs(apply(LR, 2, max)) < abs(apply(LR, 2, min))] <- -1
  LR <- LR * repmat(s, ncol(X), 1) 

  # Compute total variances
  tv <- sum(EV)

  # Create table of explained variances for unrotated and rotated components
  VT <- matrix(c(u, (100 * u) / tv, r, (100 * r) / tv), byrow = F, 
               nrow = length(u))

  # Compute the rotated factor scores coefficients
  FSCFr <- LR %*% solve(t(LR) %*% LR)
  FSCFr <- FSCFr * repmat(sqrt(diag(D)), 1, rk) # Requires the covariance matrix
  mu <- apply(X, 2, mean)
  sigma <- apply(X, 2, sd)
  Xc <- sweep(X, 2, mu)

  # Compute the rotated factor scores from normalized raw data
  FSr = matrix(0, nrow = nrow(X), ncol = ncol(X))
  for (n in 1:nrow(X)) {
    for (m in 1:rk) {
      FSr[n, m] <- sum((Xc[n, ] / sigma) * FSCFr[, m])
    }
  }
  
  # Prettify output (VT)
  factor.labels <- paste("Factor", 1:nrow(VT), sep = ".")
  header.labels <- c("Eigenvalue.Before", "Var.Explained.Before", 
                     "Eigenvalue.After", "Var.Explained.After")
  rownames(VT) <- factor.labels
  colnames(VT) <- header.labels
  
  # return the same output
  output <- list(Unrotated = LU, Rotated = LR, Factor.Scores = FSr, 
                 Variance = VT)
  output
}
